syntax = "proto3";

package database;


service DatabaseService {
  rpc AddRoom (AddRoomRequest) returns (AddRoomResponse) {}
  rpc GetRoom (GetRoomRequest) returns (GetRoomResponse) {}
  rpc GetAllRooms (GetAllRoomsRequest) returns (GetAllRoomsResponse) {}

  rpc AddDevice (AddDeviceRequest) returns (AddDeviceResponse) {}
  rpc GetDevicesInRoom (GetDevicesInRoomRequest) returns (GetDevicesInRoomResponse) {}

  rpc AddUser (AddUserRequest) returns (AddUserResponse) {}
  rpc GetUser (GetUserRequest) returns (GetUserResponse) {}
  rpc SetUserPreferences (SetUserPreferencesRequest) returns (SetUserPreferencesResponse) {}

  rpc AddUserToRoom (AddUserToRoomRequest) returns (AddUserToRoomResponse) {}
  rpc RemoveUserFromRoom (RemoveUserFromRoomRequest) returns (RemoveUserFromRoomResponse) {}
  rpc GetUsersInRoom (GetUsersInRoomRequest) returns (GetUsersInRoomResponse) {}
  rpc GetUserRoom (GetUserRoomRequest) returns (GetUserRoomResponse) {}
}

/* Basic types */

message Room {
  string id = 1;
  string name = 2;
}

message Device {
  string id = 1;
  string name = 2;
}

message Preference {
  float temperature = 1;
  int32 fan_speed = 2;
}

message User {
  string id = 1;
  string email = 2;
  optional Preference preferences = 3;
}

/* Requests and responses */

/* Rooms */

message AddRoomRequest {
  Room room = 1;
}

message AddRoomResponse {
  Room room = 1;
}

message GetRoomRequest {
  string id = 1;
}

message GetRoomResponse {
  Room room = 1;
}

message GetAllRoomsRequest {
}

message GetAllRoomsResponse {
  repeated Room rooms = 1;
}

message GetUserRoomRequest {
  string user_id = 1;
}

message GetUserRoomResponse {
  optional Room room = 1;
}

/* Devices */

message AddDeviceRequest {
  string room_id = 1;
  Device device = 2;
}

message AddDeviceResponse {
  Device device = 1;
}

message GetDevicesInRoomRequest {
  string room_id = 1;
}

message GetDevicesInRoomResponse {
  repeated Device devices = 1;
}

/* Users */

message AddUserRequest {
  User user = 1;
}

message AddUserResponse {
  User user = 1;
}

message GetUserRequest {
  string id = 1;
}

message GetUserResponse {
  User user = 1;
}

message SetUserPreferencesRequest {
  string user_id = 1;
  Preference preferences = 2;
}

message SetUserPreferencesResponse {
  Preference preferences = 1;
}

message AddUserToRoomRequest {
  string user_id = 1;
  string room_id = 2;
}

message AddUserToRoomResponse {
  string user_id = 1;
}

message RemoveUserFromRoomRequest {
  string user_id = 1;
  string room_id = 2;
}

message RemoveUserFromRoomResponse {
  string user_id = 1;
}

message GetUsersInRoomRequest {
  string room_id = 1;
}

message GetUsersInRoomResponse {
  repeated User users = 1;
}
